package.json

```diff
@@ -18,19 +18,20 @@
   "types": "lib/peg.d.ts",
   "bin": "bin/peggy",
   "scripts": {
-    "pretest": "npm run parser",
-    "test": "npm run test:bare",
-    "test:bare": "mocha --recursive test --exclude test/browser.stub.js --reporter min",
+    "clean": "rimraf -f build && rimraf -f browser && mkdir build && cd build && mkdir ts && mkdir rollup && mkdir terser && cd .. && mkdir browser",
+    "terser": "terser -c passes=2 -m --sequences 40 build/rollup/peg.umd.js > build/terser/peggy.umd.min.js && terser -c passes=2 -m --sequences 40 --module build/rollup/peg.esmodule.js > build/terser/peggy.esmodule.min.js",
+    "rollup": "rollup -c",
+    "ts": "tsc --build tsconfig.json",
+    "set_version": "node -e \"const fs = require('fs'), pkg = JSON.parse(fs.readFileSync('./package.json')); fs.writeFileSync('./lib/version.js', '\\n\\'use strict\\';\\n\\nmodule.exports = \\'' + pkg.version + '\\';\\n');\"",
+    "test": "jest --coverage",
     "lint": "eslint . --ext .js bin/peggy",
-    "benchmark": "./benchmark/run",
-    "build": "npm run browser:build && npm run benchmark:build && npm run test:build",
-    "browser:clean": "rm -rf browser/",
-    "prebrowser:build": "npm run parser",
-    "browser:build": "browserify lib/peg.js -s peggy -o browser/peggy.min.js",
-    "postbrowser:build": "node tools/header browser/peggy.min.js && cp browser/peggy.min.js docs/vendor/peggy/peggy.min.js",
-    "benchmark:build": "browserify benchmark/browser.stub.js -o docs/js/benchmark-bundle.min.js && node tools/header docs/js/benchmark-bundle.min.js",
-    "test:build": "browserify test/browser.stub.js -o docs/js/test-bundle.min.js && node tools/header docs/js/test-bundle.min.js",
-    "parser": "node bin/peggy -o lib/parser.js --format commonjs src/parser.pegjs"
+    "benchmark": "node ./benchmark/run_bench.js",
+    "parser": "node bin/peggy -o lib/parser.js --format commonjs src/parser.pegjs",
+    "distro": "rm -f browser/peggy.min.js && cp build/terser/peggy.umd.min.js browser/peggy.min.js",
+    "audit": "npm run test && npm run benchmark",
+    "make": "npm run ts && npm run rollup && npm run terser && npm run parser",
```

Comment:
I agree with this.  It would be ok to re-run all of the tools and ensure you get the same output as what was checked in.

Reply:
>I think is is dangerous to update parser during build. 

Maybe I'm misunderstanding, but, it seems like the parser is rebuilt after things it's depending on have changed, which means it's no matter what wrong, and before anything using it is run, meaning it's guaranteed in-sync

This build used to be a non-linear rat's nest, so maybe I misunderstood

&nbsp;

> If parser has changed that means that parser in the repo would differ from the one in the distributive, which is unacceptable.

I don't understand a pathway by which this can happen.

I'm not saying you're wrong.  I am saying I don't understand yet.

I'm happy to fix whatever it is, but I need clarification first please

&nbsp;

> Also, can you explain, what the ts task does? Sorry for maybe dumb question, I not familiar with TypeScript.

It runs typescript.  `tsc` means `typescript compiler`.

&nbsp;

> I agree with this. It would be ok to re-run all of the tools and ensure you get the same output as what was checked in.

If someone could baby-step me through the temporal error I'd appreciate it.  I'm missing the boat

Reply:
>I think is is dangerous to update parser during build. 

Maybe I'm misunderstanding, but, it seems like the parser is rebuilt after things it's depending on have changed, which means it's no matter what wrong, and before anything using it is run, meaning it's guaranteed in-sync

This build used to be non-linear, so maybe I misunderstood

&nbsp;

> If parser has changed that means that parser in the repo would differ from the one in the distributive, which is unacceptable.

I don't understand a pathway by which this can happen.

I'm not saying you're wrong.  I am saying I don't understand yet.

I'm happy to fix whatever it is, but I need clarification first please

&nbsp;

> Also, can you explain, what the ts task does? Sorry for maybe dumb question, I not familiar with TypeScript.

It runs typescript.  `tsc` means `typescript compiler`.

&nbsp;

> I agree with this. It would be ok to re-run all of the tools and ensure you get the same output as what was checked in.

If someone could baby-step me through the temporal error I'd appreciate it.  I'm missing the boat

Reply:
> > Also, can you explain, what the ts task does? Sorry for maybe dumb question, I not familiar with TypeScript.
>
> It runs typescript. tsc means typescript compiler.

This could be better explained.

`tsc --build foo` means "run the typescript compiler with the intent of creating a source transformation, governed by the config file `foo`," which in turn is expected to be `jsonc` in a file with the `.json` extension.

`typescript` is a double sided IR engine, meaning you can tell it what language to interpret as, and what language to produce as.

TL;DR: it scans `./lib`, interprets the JS as TS with `unknown` for every type, converts, shims, and downlevels to es6 modules, and deposits the same directory structure under `./build/ts`, where the next tool (`rollup`) will pick up the residues for its part.

Given the current configuration, what `typescript` will do is 
1. scan the tree underneath of `src` (key `include` contains an array of globs of targets,)
1. produce an end output in `es2015 michaeljacksonscript modules` (key `compilerOptions.module`,)
1. deposit output in the directory `./build/ts/` subdirectory relative to source directory under `lib` (key `compilerOptions.outDir`, key `compilerOptions.rootDir`,)
1. accept and validate javascript (key `compilerOptions.acceptJs`, key `compilerOptions.checkJs`,)
1. produce a sourcemap, a typescript declaration `.d.ts`, and a map for the `.d.ts`, (key `sourceMap`, key `declaration`, key `declarationMap`,) 
1. produce a layer of interoperability code when interacting with `commonjs` (key `compilerOptions.esModuleInterop`, will become a no-op with #127 and the collapse of the commonjs/es6 divide)
1. ignore that we don't have internal types for a lot of stuff yet (key `compilerOptions.skipLibCheck`,)
1. disallow two different includes to hit the same file from varying case, which works on mac and windows but not linux (key `compilerOptions.forceConsistentNamingInFilenames`,)
1. and strip comments (key `compilerOptions.removeComments`.)

With time, this will add a series of correctness checks emotionally similar to what `eslint` provides but based in a different kind of static analysis.  They are not yet turned on due to temporary deficiencies in the new build process, as it's being built.

Reply:
This is roughly the babel build step

Reply:
> Maybe I'm misunderstanding, but, it seems like the parser is rebuilt after things it's depending on have changed, which means it's no matter what wrong, and before anything using it is run, meaning it's guaranteed in-sync

It rebuilds before the tests are run and that mean that you will run tests with new, untested parser, that automatically mean that you can't trust the test results. This is especially clearly visible if you change the generator and did a mistake in it, as me himself while working on #118

> I don't understand a pathway by which this can happen.

For example, just speculate, that I make a PR that add some pass which download and prepend to the generated parser some, well, mining script, and that PR was merged. Repository file is not affected, but packaged file will have some untrusted code.

I believe I hear about/saw projects where special tool was checking that code in the repository corresponds code in the package.

The last argument, that with the self-modified build it is just unreproducible.

> If someone could baby-step me through the temporal error I'd appreciate it. I'm missing the boat

I'm thinking about running something like `git status` and checking the result code

> This could be better explained.

Thanks

Reply:
> It rebuilds before the tests are run and that mean that you will run tests with new, untested parser

I apologize.  I'm still confused.  Isn't ... isn't this what we want?  For the untested parser to have tests run against it?

&nbsp;

> For example, just speculate, that I make a PR that add some pass which download and prepend to the generated parser some, well, mining script, and that PR was merged. Repository file is not affected, but packaged file will have some untrusted code.

if you're saying "don't make the build system this way because someone could alter the build system and do something bad," literally no build system is ever immune to someone who can alter it

&nbsp;

> I believe I hear about/saw projects where special tool was checking that code in the repository corresponds code in the package.

I assume you mean `plutov-slack-client`.  

What `plutov-slack-client` did was to publish evil code, then to change the code in github, and leave the version number the same.

If an attacker has the ability to publish bad code, nothing in `npm scripts` will ever stop them.  This is like trying to prevent a burglar from getting into the house by locking your jewelry box.  By the time they're able to do this, it's much too late.

&nbsp;

> If someone could baby-step me through the temporal error I'd appreciate it. I'm missing the boat
>
> I'm thinking about running something like git status and checking the result code

If you're trying to prevent a difference between the package manager and github, you can't.  That just isn't how github works.

The only appropriate steps are to sign the package and consider switching to github package manager.

I don't see any device by which you can use the contents of a github repo to secure the repo against someone who can alter the repo.

This isn't something you should be attempting to invent on the spot.  These topics are well understood.

Please don't block this PR while you attempt to invent security concerns which are no different than the existing situation.  If you want to work on this, do it in a different PR.